File: 22/e03.ts
import {readFileSync} from "fs";

const TEST = false

const num = __filename.split(".")[0]
const inputFileName = TEST ? num + "-test.txt" : num + "-input.txt"

const content = readFileSync(inputFileName).toString().trim()
const lines = content.split("\n")
const char = "A"
const lowerStart = char.toLowerCase().charCodeAt(0)
const upperStart = char.charCodeAt(0)
let groups: string[][] = []

let sum = 0
let sum2 = 0
let i = 0
for (let l of lines) {
  const group = Math.floor(i / 3)
  groups[group] = [...(groups[group] || []), l]
  const part1 = l.slice(0, l.length  / 2)
  const part2 = l.slice(l.length  / 2)

  let found = ""
  for (let c of part1) {
    if(part2.includes(c) && !found.includes(c)){
      found += c
      let charCode = c.charCodeAt(0)
      if(charCode >= lowerStart) charCode += -lowerStart + 1;
      else charCode += -upperStart + 27;
      console.log(c, charCode)
      sum += charCode
    }
  }
  i++;
}

for (let g of groups ){
    let found = ""
    for (let c of g[0]) {
      if(!found.includes(c) && g[1].includes(c) && g[2].includes(c)) {
          found += c
          let charCode = c.charCodeAt(0)
          if(charCode >= lowerStart) charCode += -lowerStart + 1;
          else charCode += -upperStart + 27;
          console.log(c, charCode)
          sum2 += charCode
      }
    }
}


console.log(sum)
console.log(sum2)

File: 22/e03.nim
import strutils, sequtils, math, algorithm

template FILENAME: string = instantiationInfo().filename

const TEST = 0;

let day = FILENAME.split(".")[0];
var fn = day & "-input.txt";

if bool(TEST):
    fn = day & "-test.txt";

let contents = readFile(fn);
let fileLines = contents.strip().splitLines();

const lowerStart = int('a')
const upperStart = int('A')

var i = 0;
var 
    sum = 0
    sum2 = 0
for l in fileLines:
    var found = ""
    var 
        a = l.substr(0, int(l.len / 2) - 1)
        b = l.substr(int(l.len / 2), l.len)
    for c in a:
        if(not b.contains(c) or found.contains(c)): 
            continue
        found.add(c)
        var charCode = int(c)
        if(charCode >= lowerStart):
            charCode += -lowerStart + 1;
        else:
            charCode += -upperStart + 27;
        sum += charCode
        echo c, " ", charCode

    found = ""
    for c in l:
        if(( i + 1 ) mod 3 == 0):
            let 
                l2 = fileLines[i-1]
                l1 = fileLines[i-2]
            if(found.contains(c) or not l1.contains(c) or not l2.contains(c)):
                continue
            found.add(c)
            var charCode = int(c)
            if(charCode >= lowerStart):
                charCode += -lowerStart + 1;
            else:
                charCode += -upperStart + 27;
            sum2 += charCode

    i += 1;

echo "1: ", sum
echo "2: ", sum2
